### PR TITLE
✨Fuzzing implementation

### PR DESCRIPTION
**Pull request by antedotee**
## Summary
Implement comprehensive fuzzing test suite for KubeStellar to enhance security and identify potential bugs through automated testing with random inputs.

## Problem
KubeStellar lacked systematic fuzzing tests, leaving critical components vulnerable to undiscovered bugs, crashes, and security vulnerabilities. Without automated fuzzing, edge cases in JSONPath parsing, CRD validation, label parsing, and API group handling could go undetected, potentially leading to runtime failures, security exploits, or unexpected behavior in production environments. This gap in testing coverage left the project exposed to input validation vulnerabilities and didn't meet modern security testing standards for enterprise-grade software.

## Solution
Implemented a comprehensive fuzzing test suite using Go's built-in fuzzing support (Go 1.18+) to systematically test KubeStellar components with random, unexpected, and malformed inputs, ensuring robust error handling and security validation.

**Key features:**
- Complete coverage of critical KubeStellar components (JSONPath, CRD validation, label parsing, API groups)
- Integration with Go's native fuzzing framework for efficient testing
- OSS-Fuzz integration for continuous security testing
- Comprehensive seed corpus for guided fuzzing
- Detailed documentation and best practices

**Results:**
- Establishes robust security testing for KubeStellar components
- Identifies potential bugs and vulnerabilities before production deployment
- Improves input validation and error handling across the codebase
- Meets enterprise security testing standards

Relates to #2626

### CODE REVIEW

**Comment by antedotee** [comment]
I ran test on my local and got this:
image

I think there will still be a lot of improvements since I have implemented basic fuzz tests.

**Comment by antedotee** [comment]
I have implemented fuzzing in the CI pipeline but the OSSFuzz requires the project to be listed. The best alternative perhaps there is OSSFuzz(also used by other large projects like KubeEdge etc., most other ci fuzz tools are probably paid). We will have to list kubestellar as a project on OSSFuzz and then the fuzz integration will work as intended. To do that, we will have to raise a pull request on ossfuzz github repo containing Dockerfile, build.sh and project.yml. I think the project maintainer or owner should raise it, or I can also raise it by closely communicating with maintainer or owner.
I would be happy to discuss what are the other alternatives to carry fuzzing implementation and implement the best solution possible. If we dont wish to use OSSFuzz, we can use **clusterfuzzlite**.
Here is the attached link on accepting new projects: https://google.github.io/oss-fuzz/getting-started/accepting-new-projects/
I am also attaching a merged PR on how they raised their PR(just for better reference): https://github.com/google/oss-fuzz/pull/7755
Thanks and regards.

**Comment by antedotee** [comment]
@clubanderson @MikeSpreitzer Can you please take a look into it😅(Some checks are failing, I am fixing them meanwhile, but the bigger question is what service should be used in future)?

**Comment by clubanderson** [comment]
/ok-to-test

@antedotee are you going to continue on this PR?

**Comment by MikeSpreitzer** [comment]
@antedotee: The reason for the action hash checking is explained in https://docs.kubestellar.io/unreleased-development/contribution-guidelines/contributing-inc/#github-action-reference-discipline . See `hack/gha-reversemap.sh` for a script for doing updates and checks using this discipline.

**Comment by MikeSpreitzer** [comment]
@antedotee: I am not familiar with the fuzzing tool used here, you will have to debug the failure in your new workflow yourself.

**Comment by antedotee** [comment]
I have used native go fuzzing for now to implement fuzzing.
Waiting for the review and if needed, to implement the suggestions given.
Thanks and regards.

**Comment by katara-Jayprakash** [comment]
we don't need to build these things into GitHub Action , in kubesteller we use prow . and we can integrate them in prow too. and ya I think this approach is good. but make sure  we  got all the things that we required to pass the  openssf practises @MikeSpreitzer any other insights sir. 
 @antedotee

**Comment by MikeSpreitzer** [comment]
Yes, we prefer using Prow over GitHub Actions workflows.

File: .github/workflows/cifuzz.yml [hunk 1]
```diff
@@ -0,0 +1,108 @@
...
+    runs-on: ubuntu-22.04
+    
+    steps:
+      - name: Checkout code
+        uses: actions/checkout@11bd71901bbe5b1630ceea73d27597364c9af683
+
+      - name: Set up Go
+        uses: actions/setup-go@d35c59abb061a4a6fb18e82ac0862c26744d6ab5
+        with:
+          go-version: 1.22.x
```
Comment (MikeSpreitzer) [inline .github/workflows/cifuzz.yml:35]: Minor version 22 of Go is pretty outdated. We should be using 24 or 25 now.

The ".x" looks odd. It is not documented as acceptable in https://github.com/actions/setup-go#supported-version-syntax . In common parlance, ".x" is used to indicate uncertainty. I think that we want certainty.

The documentation for this action mentions some other possibilities too. I think that the most appropriate would be to use the latest stable version of Go.

File: hack/gha-reversemap.sh
Comment (MikeSpreitzer) [inline hack/gha-reversemap.sh:1]: The yq version problem has been solved in `main`. Please rebase this PR's branch onto `main`.

File: test/fuzz/verify-setup.sh [hunk 2]
```diff
@@ -0,0 +1,180 @@
...
+NC='\033[0m' # No Color
+
+echo "🔧 KubeStellar Fuzz Testing Setup Verification"
+echo "=============================================="
+
+# Check Go version
+echo -n "Checking Go version... "
+if command -v go &> /dev/null; then
+    GO_VERSION=$(go version | cut -d' ' -f3)
+    echo -e "${GREEN}✓${NC} Found Go $GO_VERSION"
```
Comment (MikeSpreitzer) [inline test/fuzz/verify-setup.sh:34]: It is better to do the following, because it makes less stuff subject to interpretation.

```shell
GREEN=$'\033[0;32m'
NC=$'\033[0m' # No Color
    echo "${GREEN}✓${NC} Found Go $GO_VERSION"
```

File: test/fuzz/verify-setup.sh [hunk 3]
```diff
@@ -0,0 +1,180 @@
...
+    # Check if Go version is 1.18 or higher (required for fuzzing)
+    GO_MAJOR=$(echo $GO_VERSION | sed 's/go//' | cut -d'.' -f1)
+    GO_MINOR=$(echo $GO_VERSION | sed 's/go//' | cut -d'.' -f2)
+    
+    if [ "$GO_MAJOR" -gt 1 ] || [ "$GO_MAJOR" -eq 1 -a "$GO_MINOR" -ge 18 ]; then
+        echo -e "  ${GREEN}✓${NC} Go version supports fuzzing (requires Go 1.18+)"
+    else
+        echo -e "  ${RED}✗${NC} Go version is too old. Fuzzing requires Go 1.18 or higher"
+        exit 1
+    fi
```
Comment (MikeSpreitzer) [inline test/fuzz/verify-setup.sh:45]: This test is unnecessary. Elsewhere we are specifying a Go release much higher than 1.18.

File: test/fuzz/verify-setup.sh [hunk 4]
```diff
@@ -0,0 +1,180 @@
...
+
+# Check if we're in the right directory
+echo -n "Checking project structure... "
+if [ -f "../../go.mod" ] && grep -q "github.com/kubestellar/kubestellar" "../../go.mod"; then
+    echo -e "${GREEN}✓${NC} In KubeStellar project directory"
+else
+    echo -e "${RED}✗${NC} Not in KubeStellar project root or invalid structure"
+    echo "Please run this script from the test/fuzz directory"
+    exit 1
+fi
```
Comment (MikeSpreitzer) [inline test/fuzz/verify-setup.sh:61]: Rather than insist on the cwd of the caller, just have this script `cd` to the right directory.

File: test/fuzz/verify-setup.sh [hunk 5]
```diff
@@ -0,0 +1,180 @@
...
+    echo -e "${RED}✗${NC} Fuzz tests fail to compile"
+    echo "Run 'go test -c .' to see compilation errors"
+    exit 1
+fi
+
+echo ""
+
+# Run a quick fuzz test
+echo "🧪 Running quick fuzz test (5 seconds)..."
+echo "========================================="
```
Comment (MikeSpreitzer) [inline test/fuzz/verify-setup.sh:120]: This quick test brings a lot of stuff to maintain here, for very little practical benefit. I would delete it.

File: test/fuzz/kubestellar_fuzz_test.go [hunk 6]
```diff
@@ -0,0 +1,403 @@
...
+			// This is valid when the JSON only specifies the type
+		case v1alpha1.TypeObject:
+			// Object type can have nil Object field (omitted from JSON)
+			// This is valid when the JSON only specifies the type
+		case v1alpha1.TypeArray:
+			// Array type can have nil Array field (omitted from JSON)
+			// This is valid when the JSON only specifies the type
+		case v1alpha1.TypeNull:
+			// Null type doesn't need any fields
+		}
```
Comment (MikeSpreitzer) [inline test/fuzz/kubestellar_fuzz_test.go:401]: This `switch` statement does nothing in every case. The whole thing is a no-op, and as such it should be removed.

File: test/fuzz/kubestellar_fuzz_test.go [hunk 7]
```diff
@@ -0,0 +1,403 @@
...
+		`{"type": "Object", "object": {"key": "value"}}`,
+		`{"type": "Array", "array": [1, 2, 3]}`,
+		`{"type": "Null"}`,
+	}
+
+	for _, seed := range seedCorpus {
+		f.Add(seed)
+	}
+
+	f.Fuzz(func(t *testing.T, jsonStr string) {
```
Comment (MikeSpreitzer) [inline test/fuzz/kubestellar_fuzz_test.go:345]: The function under test never fails. This whole `f.Fuzz` is a no-op, and as such should be removed.

File: test/fuzz/kubestellar_fuzz_test.go [hunk 8]
```diff
@@ -0,0 +1,403 @@
...
+		// (though they might be nil if the path doesn't exist)
+		if len(query) > 0 && len(results) == 0 {
+			// This might indicate an issue, but not necessarily a bug
+			// as the path might not exist in our test object
+		}
+	})
+}
+
+// FuzzCRDValidation tests CRD validation with various YAML inputs
+func FuzzCRDValidation(f *testing.F) {
```
Comment (MikeSpreitzer) [inline test/fuzz/kubestellar_fuzz_test.go:113]: The name of this function is wrong. It does not test validation of custom resource definitions. Rather, it exercises parsing of custom resources.

File: test/fuzz/kubestellar_fuzz_test.go [hunk 9]
```diff
@@ -0,0 +1,403 @@
...
+`
+
+	// Add seed corpus
+	f.Add([]byte(baseBindingPolicy))
+	f.Add([]byte(baseStatusCollector))
+	f.Add([]byte(baseBinding))
+	f.Add([]byte(baseBindingPolicy + "---\n" + baseStatusCollector))
+	f.Add([]byte(baseBindingPolicy + "---\n" + baseBinding))
+
+	f.Fuzz(func(t *testing.T, yamlData []byte) {
```
Comment (MikeSpreitzer) [inline test/fuzz/kubestellar_fuzz_test.go:165]: This test is pointless, it never fails.

File: test/fuzz/kubestellar_fuzz_test.go [hunk 10]
```diff
@@ -0,0 +1,403 @@
...
+		"$.spec.template.metadata.labels",
+		"$.spec.ports[0].port",
+		"$.spec.rules[0].http.paths[0].path",
+	}
+
+	for _, seed := range seedCorpus {
+		f.Add(seed)
+	}
+
+	f.Fuzz(func(t *testing.T, queryStr string) {
```
Comment (MikeSpreitzer) [inline test/fuzz/kubestellar_fuzz_test.go:54]: This test never fails. It is a no-op, and as such should be removed.

File: test/fuzz/kubestellar_fuzz_test.go [hunk 11]
```diff
@@ -0,0 +1,403 @@
...
+		reconstructed := fmt.Sprintf("%s=%s", label.Key, label.Value)
+		if reconstructed != labelStr {
+			// This might indicate an issue with parsing/formatting
+			t.Errorf("Label reconstruction mismatch: expected %q, got %q", labelStr, reconstructed)
+		}
+	})
+}
+
+// FuzzAPIGroupParsing tests API group parsing functionality
+func FuzzAPIGroupParsing(f *testing.F) {
```
Comment (MikeSpreitzer) [inline test/fuzz/kubestellar_fuzz_test.go:279]: The name of this function is wrong. It uses the singular form "APIGroup", but the function being tested parses a plural and has "APIGroups" in its name.

File: test/fuzz/kubestellar_fuzz_test.go [hunk 12]
```diff
@@ -0,0 +1,403 @@
...
+
+	f.Fuzz(func(t *testing.T, apiGroupsStr string) {
+		// Parse API groups
+		apiGroups := util.ParseAPIGroupsString(apiGroupsStr)
+
+		// Test that required groups are always included
+		if apiGroups != nil {
+			// Check that KubeStellar control group is included
+			if !apiGroups.Has("control.kubestellar.io") {
+				t.Errorf("Required API group 'control.kubestellar.io' not found in parsed groups")
```
Comment (MikeSpreitzer) [inline test/fuzz/kubestellar_fuzz_test.go:306]: The fuzzing does not guarantee that `apiGroupsStr` includes `control.kubestellar.io`, right?

File: .github/workflows/cifuzz.yml [hunk 13]
```diff
@@ -0,0 +1,108 @@
...
+          cache: true
+
+      - name: Download dependencies
+        run: go mod download
+
+      - name: Verify fuzz tests exist
+        run: |
+          echo "Checking for fuzz tests..."
+          cd test/fuzz
+          go test -list='Fuzz.*' | grep -E '^Fuzz' || echo "No fuzz tests found"
```
Comment (MikeSpreitzer) [inline .github/workflows/cifuzz.yml:45]: What's the point of including `.*` in the regular expression that is sought as a substring?

Comment (MikeSpreitzer) [inline .github/workflows/cifuzz.yml:45]: Wouldn't it be simpler to `go test -list='^Fuzz'` and omit the grep?

File: Makefile [hunk 14]
```diff
@@ -171,6 +171,23 @@ vet: ## Run go vet against code.
...
 	go test ./api/... ./cmd/... ./pkg/... -coverprofile cover.out
 
+# Fuzzing configuration
+FUZZ_TIME ?= 10s
+
+.PHONY: test-fuzz
+test-fuzz: ## Run all fuzzers in the test/fuzz folder one by one
+	@echo "Running fuzz tests..."
+	@for dir in $$(go list -f '{{.Dir}}' ./test/fuzz/...); do \
+		for test in $$(go test -list=Fuzz.* $$dir | grep ^Fuzz); do \
```
Comment (MikeSpreitzer) [inline Makefile:181]: Why not do the filtering in the r.e.?

File: test/fuzz/kubestellar_fuzz_test.go [hunk 15]
```diff
@@ -0,0 +1,403 @@
...
+	f.Fuzz(func(t *testing.T, labelStr string) {
+		// Skip empty strings
+		if strings.TrimSpace(labelStr) == "" {
+			return
+		}
+
+		// Try to parse the label
+		label, err := util.SplitLabelKeyAndValue(labelStr)
+		if err != nil {
+			// Expected for invalid label formats, just return
```
Comment (MikeSpreitzer) [inline test/fuzz/kubestellar_fuzz_test.go:260]: The test func should succeed only if either (a) the input is invalid and `util.SplitLabelKeyAndValue` returns an error or (b) the input is valid and `util.SplitLabelKeyAndValue` does NOT return an error.

File: test/fuzz/kubestellar_fuzz_test.go [hunk 16]
```diff
@@ -0,0 +1,403 @@
...
+
+// FuzzCRDValidation tests CRD validation with various YAML inputs
+func FuzzCRDValidation(f *testing.F) {
+	// Add seed corpus for CRD validation
+	baseBindingPolicy := `apiVersion: control.kubestellar.io/v1alpha1
+kind: BindingPolicy
+metadata:
+  name: test-binding-policy
+  namespace: default
+spec:
```
Comment (MikeSpreitzer) [inline test/fuzz/kubestellar_fuzz_test.go:120]: Why no `clusterSelectors` ?

File: test/fuzz/kubestellar_fuzz_test.go [hunk 17]
```diff
@@ -0,0 +1,403 @@
...
+`
+
+	baseBinding := `apiVersion: control.kubestellar.io/v1alpha1
+kind: Binding
+metadata:
+  name: test-binding
+  namespace: default
+spec:
+  bindingPolicyRef: test-binding-policy
+  workloadSelector:
```
Comment (MikeSpreitzer) [inline test/fuzz/kubestellar_fuzz_test.go:151]: These are not the fields of a `BindingSpec`.

File: test/fuzz/testdata/binding.yaml [hunk 18]
```diff
@@ -0,0 +1,13 @@
+apiVersion: control.kubestellar.io/v1alpha1
+kind: Binding
+metadata:
+  name: test-binding
+  namespace: default
+spec:
+  bindingPolicyRef: test-binding-policy
+  workloadSelector:
```
Comment (MikeSpreitzer) [inline test/fuzz/testdata/binding.yaml:8]: These are not the fields of a `BindingSpec`.

File: test/fuzz/testdata/binding_policy.yaml [hunk 19]
```diff
@@ -0,0 +1,25 @@
...
+      namespaces: ["default", "kube-system"]
+      objectSelectors:
+        - matchLabels:
+            app: test
+            tier: frontend
+        - matchExpressions:
+            - key: environment
+              operator: In
+              values: ["production", "staging"]
+  upsync:
```
Comment (MikeSpreitzer) [inline test/fuzz/testdata/binding_policy.yaml:19]: A `BindingPolicySpec` does not have an `upsync` field.

**Comment by MikeSpreitzer** [comment]
https://github.com/kubestellar/kubestellar/issues/2626#issuecomment-3262633654

File: test/fuzz/testdata/binding.yaml
Comment (MikeSpreitzer) [inline test/fuzz/testdata/binding.yaml:1]: What uses this file?

File: test/fuzz/kubestellar_fuzz_test.go (hunk 9 above)
Comment (MikeSpreitzer) [inline test/fuzz/kubestellar_fuzz_test.go:165]: This test does not test anything in this repository. It exercises things in the go standard library and in Kubernetes.

**Comment by katara-Jayprakash** [comment]
@MikeSpreitzer should I take over this and implement this with prow. I already created an issue regarding to this.

**Comment by antedotee** [comment]
Hi, @katara-Jayprakash , I was busy with my work since past week. If I am unable to resolve today, I will close this PR. Till then, I think you can pick other issues to work upon.
Thanks and regards.

**Comment by MikeSpreitzer** [comment]
I think that the author intends us to pursue #3327 instead.
